specification/flag-evaluation/hooks.md

```diff
@@ -2,4 +2,126 @@
+
+> Hook context **SHOULD** provide: provider, client
+
+> flag key, flag type, default value properties **MUST** be immutable.
+
+> The evaluation context **MUST** be mutable only within the `before` hook.
+
+### State
+
+> State **MUST** be a map of objects.
+
+> Condition: State **MUST** be immutable, if the implementation language supports a mechanism for marking data as immutable.
+
+### Hook creation and parameters
+
+> Hooks **MUST** specify at least one stage.
+
+> The `before` stage *MUST* run before flag evaluation occurs. It accepts a `hook context` (required) and `state` (optional) as parameters and returns either a `HookContext` or nothing.
+
+```
+HookContext|void before(HookContext, State)
+```
+
+> The `after` stage **MUST** run after flag evaluation occurs. It accepts a `hook context` (required), `flag evaluation details` (required) and `state` (optional). It has no return value.
+
+> The `error` hook runs when errors are encountered in the `before` or `after` state or when flag evaluation errors. It accepts `hook context` (required), `exception` for what went wrong (required), and `state` (optional). It has no return value.
+
+> The `finally` hook **MUST** run after the `before`, `after`, and `error` stages. It accepts a `hook context` (required) and `state` (optional). There is no return value.
+
+> Condition: If `finally` is a reserved word in the language, `afterAll` should be used.
+
+### Hook registration & ordering
+
+> The API, Client and invocation **MUST** have a method for registering hooks which accepts `flag evaluation options`
+
+```js
+OpenFeature.addHooks(new Hook1());
+
+//...
+
+Client client = OpenFeature.getClient();
+client.addHooks(new Hook2());
+
+//...
+
+client.getValue('my-flag', 'defaultValue', new Hook3());
+```
+
+> Hooks **MUST** be evaluated in the following order: API before, client before, invocation before, invocation after, client after, API after, invocation finally, client finally, API finally. If errors are present, the order is: invocation error, client error, API error.
+
+```
+// here we are using the order defined above (note if it wasn't defined, we'd have to order them somehow anyway
+Hooks[] allHooks = [ ...OpenFeature.getHooks(), ...this.getHooks(), ...evaluationOptions.hooks ]
+
+try {
+  // evaluation all our before hooks
+  for (h in allHooks) {
+    h.before(...);
+  }
+
+  provider.resolveValue(...);
+
+  // evaluation all our after hooks
+  for (h in allHooks.reverse()) {
+    h.after(...);
+  }
+} catch (Error err) {
+  // evaluation all our error hooks
+  for (h in allHooks.reverse()) {
+    h.error(...);
+  }
+} finally {
+  // evaluation all our finally hooks
+  for (h in allHooks.reverse()) {
+    h.finally(...);
+  }
+}
+```
+
+> When an error occurs in hook evaluation, further evaluation **MUST** stop and error hook evaluation begins.
```

Comment:
in this section I was a bit confused about when we were talking about errors in the provider evaluation vs errors in the hook evaluation. It might be clearer if the language was a bit more specific. For example, rather than saying "if an error is encountered in the error stage" we could say "if an error is encountered during evaluation of a hook in the error stage", or just "during evaluation of an error hook"

Reply:
> If an error occurs during the evaluation of `before` or `after` hooks, any remaining hooks in the `before` or `after` stages **MUST** not be invoked.

^ I'm going to try that. The intent is errors are kinda like an exception in that it stops the flow and bubbles up to the error stage.